### PR TITLE
Add is_public to groups table to allow for private groups

### PR DESCRIPTION
**Pull request by lukebarnard1**
Prevent group API access to non-members for private groups

Also make all the group code paths consistent with `requester_user_id` always being the User ID of the requesting user.

### CODE REVIEW

File: synapse/groups/groups_server.py [hunk 1]
```diff
@@ -67,6 +67,10 @@ def check_group_is_ours(self, group_id, and_exists=False, and_is_admin=None):
         if and_exists and not group:
             raise SynapseError(404, "Unknown group")
 
+        is_user_in_group = yield self.store.is_user_in_group(requester_user_id, group_id)
+        if is_user_in_group or not group.is_public:
+            raise SynapseError(404, "Unknown group")
```
Comment (erikjohnston) [inline synapse/groups/groups_server.py]: We should probably only do this if the group exists, in particular this currently doesn't work well with create_group. (Speaking of which, you forgot to add requester_user_id to create_group, which is why its failing)

File: synapse/storage/schema/delta/46/group_server.sql [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- whether non-members can access group APIs
+ALTER TABLE groups ADD COLUMN is_public BOOL DEFAULT 1 NOT NULL;
```
Comment (erikjohnston) [inline synapse/storage/schema/delta/46/group_server.sql]: You'll need to change the schema version in synapse.storage.prepare_database

Comment (erikjohnston) [reply synapse/storage/schema/delta/46/group_server.sql]: You'll need to add this to the scripts/synapse_port_db list of tables with boolean columns

File: synapse/groups/groups_server.py (hunk 1 above)
Comment (turt2live) [reply synapse/groups/groups_server.py]: Shouldn't that also be `if not is_user_in_group...`?

Comment (lukebarnard1) [reply synapse/groups/groups_server.py]: yes o\

**Comment by erikjohnston** [comment]
(Ideally we should also add a sytest for this)

**Pull request merged**